library/psa_crypto_core.h

```diff
@@ -111,6 +117,22 @@ static inline int psa_is_key_slot_locked(const psa_key_slot_t *slot)
     return slot->lock_count > 0;
 }
 
+/** Test whether a key slot is locked for writing.
```

Comment:
As I said on the call today, I'm not convinced that the functions added here are the ones we need. (But maybe they are, I'm just not sure.) It seems to me that what we really need is primitives that do state transitions, not primitives that specifically act on the combined reader count and exclusivity flag. We've been debating what error code to return from those functions and maybe the problem is that these functions don't have enough information to decide.

How about trying to reason like this. A slot has several states:

* unused — no information in it, nobody trying to use it. Anyone can come and grab it.
* filling — one thread is currently loading or creating material to fill it. That thread is responsible for the next state transition.
* operational — the slot contains a key and any thread can start using the key (after getting registered).
* throttled — the slot contains a key, but only for the sake of threads that are already registered to use it. Newcomers should pretend the key in this slot doesn't exist. (I'm not sure if this state exists in the MVP.)
* destroying — the slot has no keys using it, and one thread is currently cleaning it up. That thread is responsible for the next state transition.

Then we need three functions:

* `state_transition(new_state)` → `SUCCESS`, or `BAD_STATE` if the transition is not possible at this time, or `CORRUPTION_DETECTED` if the transition doesn't make sense at all.
* `register_read` — succeeds only in the “operational” state, increases the reader count.
* `unregister_read` — succeeds only in the “operational” state, decreases the reader count.

This puts the responsibility for state transition and reader count in the same place, which I think makes reasoning easier.

Reply:
`unregister_read` will need to succeed in the `throttled` state as well.

(It exists in the MVP, that is how we delete keys: it becomes logically deleted and the last reading thread does the actual deletion.)

Reply:
(semantics) can we call it "pending delete" rather than "throttled"? "Throttling" has connotations of limiting bandwidth, but this is more of a drain

Reply:
For `psa_get_empty_key_slot` we need to be able to check whether an occupied state is currently being read or not (i.e. reader count != 0). Does this mean that we need two distinct states: {operational with readers , operational without readers}?

Reply:
> can we call it "pending delete" rather than "throttled"?

Sure, that's a better name.

> we need to be able to check whether an occupied state is currently being read or not (i.e. reader count != 0). Does this mean that we need two distinct states: {operational with readers , operational without readers}?

I think this is best handled by a fourth function `has_readers`, or even `is_empty_slot`.

Reply:
> I think this is best handled by a fourth function has_readers, or even is_empty_slot

`has_readers` is probably better as there is difference between an occupied slot that doesn't have any readers and a slot that is actually empty.

Reply:
I have uploaded a pr for changing the thread-safety document to implement a new slot state approach here #8618. I don't think a destroying state is necessary, but I am happy to be proven wrong :)